### ISSUE TITLE
Zarr3 vs Zarr2 benchmarking

### ISSUE THREAD

**Issue by anaprietonem** [ecmwf/anemoi-datasets#486]
Initial tests suggest that model performance is not impacted (in terms of loss function or skills)  but we do see a slight drop in throughput and also increase in CPU memory usage. We want to run some bencharmking run to better understand that.

**Comment by frazane** [comment]
Just for background: https://github.com/zarrs/zarr_benchmarks 

It might be interesting to consider [`zarrs-python`](https://github.com/zarrs/zarrs-python).

**Comment by anaprietonem** [comment]
Hello! Thanks @frazane. @cathalobrien and @floriankrb have been doing some bencharmking, I include some the results and summary below.

The plots report a comparison of throughput performance for datasets in both o96 and n320 resolution using zarr2 vs zarr3 python package. Note the datasets mentioned are zarr2 datasets - what we are interested in here is checking performance degradation as we already have a large number of datasets that we don't want to recreate when migrating. 
All benchmarks have been done in ATOS.O96 plots check 2 different filesystems - ATOS standard filesystem and  ATOS fast SSD file system.
 
Zarr3 includes the option of parallelisation using multi-threading, Florian noticed this was suboptimal and looked at replacing it with multi-processes. 'Torch' refers to an implementation that includes instantiation and fetching  through a Pytorch Dataloader, while thread and multi-processes refer to calling open_dataset and reading random chunks out of it. 

<img width="1596" height="747" alt="Image" src="https://github.com/user-attachments/assets/cb7db4c1-419e-4534-af91-43e295772eaa" />
<img width="1596" height="747" alt="Image" src="https://github.com/user-attachments/assets/01998d84-79af-4c33-8b9f-fc03605edfb8" />
<img width="1596" height="747" alt="Image" src="https://github.com/user-attachments/assets/7df2ab96-7206-4b99-b964-04d45931f50b" />


Throughput measured via anemoi-training (n320 examples) - code run by Cathal!

Stretched grid

Plain Text
--------------------
avThroughputIterPerS: 1.10iter/s (date: 2026-01-19, commit: 2385b9debbd5f38673309574aa3870e2531f9f39)
peakMemoryMB: 12721.03MB (date: 2026-01-19, commit: 2385b9debbd5f38673309574aa3870e2531f9f39)
--------------------
(Server location: '/XXXXX')
INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:716 Local benchmark results:
--------------------
avThroughputIterPerS: 0.79iter/s (date: 2026-01-30, commit: c71da4968f98b4ceb7b21877f955331ed207e117)
peakMemoryMB: 12721.03MB (date: 2026-01-30, commit: c71da4968f98b4ceb7b21877f955331ed207e117)
--------------------

INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:719 Comparing local benchmark results against reference values from the server
INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:325 FAIL. Local value for avThroughputIterPerS has degraded compared to the reference value (0.79iter/s local vs 1.10iter/s reference)
INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:325 PASS. Local value for peakMemoryMB is within 1% tolerance of the reference value (12721.03MB local vs 12721.03MB reference)
INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:732 Profiling artifacts from failed run stored under: stretched_artifacts.tar.gz

LAM 
----
Plain Text
--------------------
avThroughputIterPerS: 0.95iter/s (date: 2026-01-23, commit: bc64bac8df31f8ffb9951de4f7d02bf4f442a4be)
peakMemoryMB: 12227.68MB (date: 2026-01-23, commit: bc64bac8df31f8ffb9951de4f7d02bf4f442a4be)
--------------------
(Server location: 'XXXXX')
INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:716 Local benchmark results:
--------------------
avThroughputIterPerS: 0.62iter/s (date: 2026-01-30, commit: c71da4968f98b4ceb7b21877f955331ed207e117)
peakMemoryMB: 12227.68MB (date: 2026-01-30, commit: c71da4968f98b4ceb7b21877f955331ed207e117)
--------------------

INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:719 Comparing local benchmark results against reference values from the server
INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:325 FAIL. Local value for avThroughputIterPerS has degraded compared to the reference value (0.62iter/s local vs 0.95iter/s reference)
INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:325 PASS. Local value for peakMemoryMB has improved compared to the reference value (12227.68MB local vs 12227.68MB reference)
INFO     anemoi.training.diagnostics.benchmark_server:benchmark_server.py:732 Profiling artifacts from failed run stored under: lam_artifacts.tar.gz
----


We are still trying to get a few more data points. But above results suggest a potential degradation of throughput specially for datasets at coarse resolution. This gap might depend on model's size and time it takes for model's training step as it might be that part is then slow and hence the degradation in throughput when reading data  is damped. 

As discussed with @ecmwf/anemoi_technical_subgroup - decision is to go ahead and still include the code for zarr3 in main in anemoi-datasets but do not remove the pyproject constrain on zarr2. This way if people want's to experiment and force getting zarr3 they can do it easily, Anemoi-datasets already support zarr3, and in the future if we find a way to optimise this further we can just simply update the pyproject.

**Comment by b8raoult** [comment]
My hunch is that the issue comes from `asyncio`, as it is the main difference between Zarr 2 and 3. Below is a profile of a very simple code, looping through a Zarr array.

Zarr I/O thread:

<img width="1027" height="572" alt="Image" src="https://github.com/user-attachments/assets/bc203991-a5d5-44a7-a76c-5dabc5fd6f39" />

One of the asynio thread:

<img width="891" height="395" alt="Image" src="https://github.com/user-attachments/assets/182222e8-2f04-4401-8019-f340b3a4a531" />

So, my gut feeling is that the I/O are so fast on lustre that the overhead of the asyncio framework (threads, sendto, select, etc) is non negligible. We still need to understand why  the workers are waiting between calls.

**Comment by TomNicholas** [comment]
@b8raoult have you tried out https://github.com/zarr-developers/zarr-python/pull/3715? @d-v-b made that PR specifically to test the same theory you have!

**Comment by floriankrb** [comment]
This should not be closed yet
